includes/ConvertToBlocks/MigrationAgent.php

```diff
@@ -0,0 +1,215 @@
+		update_option( 'ctb_posts_to_update', $posts_to_update );
+		update_option( 'ctb_cursor', -1 );
+
+		return $this->next();
+	}
+
+	/**
+	 * Stops the batch conversion if running and resets the previous session.
+	 */
+	public function stop() {
+		update_option( 'ctb_running', 0 );
+		update_option( 'ctb_posts_to_update', [] );
+		update_option( 'ctb_cursor', -1 );
+	}
+
+	/**
+	 * Returns the current status of the batch conversion.
+	 *
+	 * @return array
+	 */
+	public function get_status() {
+		$running         = get_option( 'ctb_running' );
+		$posts_to_update = get_option( 'ctb_posts_to_update' );
+
+		if ( empty( $posts_to_update ) ) {
+			$posts_to_update = [];
+		}
+
+		$total  = count( $posts_to_update );
+		$cursor = get_option( 'ctb_cursor' );
+
+		if ( $total > 0 ) {
+			$progress = round( ( $cursor + 1 ) / $total * 100 );
+		} else {
+			$progress = 0;
+		}
+
+		return [
+			'running'  => $running,
+			'cursor'   => $cursor,
+			'total'    => $total,
+			'progress' => $progress,
+			'active'   => $this->get_client_link( $posts_to_update[ $cursor ] ?? 0 ),
+		];
+	}
+
+	/**
+	 * Returns a boolean based on whether a migration is currently running.
+	 *
+	 * @return bool
+	 */
+	public function is_running() {
+		$running = get_option( 'ctb_running' );
+		return ! empty( $running );
+	}
+
+	/**
+	 * Updates the progress cursor to jump to the next post in the queue.
+	 */
+	public function next() {
+		$posts_to_update = get_option( 'ctb_posts_to_update' );
+		$total           = count( $posts_to_update );
+		$cursor          = get_option( 'ctb_cursor' );
+
+		if ( $cursor + 1 < $total ) {
+			$next_cursor = ++$cursor;
+			update_option( 'ctb_cursor', $next_cursor );
+
+			return $this->get_client_link( $posts_to_update[ $next_cursor ] );
+		} elseif ( $cursor + 1 === $total ) {
+			update_option( 'ctb_running', 0 );
+			return false;
+		} else {
+			return false;
+		}
+	}
+
+	/**
+	 * Returns the next post URL to migrate.
+	 *
+	 * @param int $post_id The next post id.
+	 * @return string
+	 */
+	public function get_client_link( $post_id ) {
+		if ( empty( $post_id ) ) {
+			return '';
+		}
+
+		$edit_post_link = admin_url( 'post.php' );
+
+		$args = [
+			'post'       => $post_id,
+			'action'     => 'edit',
+			'ctb_client' => $post_id,
+		];
+
+		return add_query_arg( $args, $edit_post_link );
+	}
+
+	/**
+	 * Returns the list of post ids that need to be migrated.
+	 *
+	 * @param array $opts Optional opts
+	 * @return array
+	 */
+	public function get_posts_to_update( $opts = [] ) {
+		if ( ! empty( $opts['post_type'] ) ) {
+			$post_type = explode( ',', $opts['post_type'] );
+			$post_type = array_filter( $post_type );
+
+			if ( empty( $post_type ) ) {
+				$post_type = [ 'post', 'page' ];
+			}
+		} else {
+			$post_type = [ 'post', 'page' ];
+		}
+
+		$query_params = [
+			'post_type'           => $post_type,
+			'post_status'         => 'publish',
+			'fields'              => 'ids',
+			'posts_per_page'      => -1,
```

Comment:
I think this would need to be updated to be a paginated loop.

Reply:
@dsawardekar if you agree on ^ this, mind a quick update to make that update?

Reply:
@gthayer @jeffpaul We do need the `posts_per_page=-1` here because of the way this has to be structured. The WP CLI seeds the loop with the `ids` query and runs only once when the WP CLI command is started. Once the agent link is opened, the loop runs entirely in the browser via redirects.